doc/developer/design/2022-04-27_platform_upgrade_test.md

```diff
@@ -0,0 +1,158 @@
+
+## Scenario
+
+A **Scenario** is a sequence of steps by which the **Initial State** will get converted into a **Final State**
+
+## Steps
+
+A **Step** in a scenario is a single action that is being undertaken with an end goal of performing the complete upgrade once all **Step**s have been completed. A **Step** can be one of the following:
+
+- a non-Kubernetes **Action** that operates on the cluster, e.g. starting or stopping a Mz process in a non-Kuberenetes environment;
+- a **KubernetesCommand** that is sent to Kubernetes API which results in some operation being carried by K8s against the cluster. For example, instructing K8s to tear down a Pod would cause Kubernetes to kill a particular set of running containers;
+- a Kubernetes action that Kubernetes performs on its own, such as restarting containers that K8s is instructed to keep up at all times
+- a **MaterializeCommand** that is sent to Mz directly, most often via the SQL interface, to manage Materialize. This would usually be a testdrive fragment
+
+# Checks
+
+In addition to the steps above, the framework will perform the following during the process:
+- **ImplicitCheck**s will be automatically executed at every stage of the test without the need to be explicitly scripted in and without relying on specific database objects to have been created or present;
+- **ExplicitCheck**s will run at the pre-defined important times defined in the scenario
+
+# Explicit Checks
+
+Explicit checks refers to creating and populating custom database objects with the desire to explicitly validate their contents later in the upgrade scenario.
+
+Explicit checks have the ability to decide by themselves if they can run successfully in a particular upgrade context.
+
+## Defining an Explicit Check
+
+Each Check is a Python class that has the following methods:
+
+```
+class ExplicitCheck:
+    def populate(self): -> List[Td]:
+       ...
+
+    def validate(self): -> Td:
+       ...
+```
+
+`populate()` returns a list of testdrive fragments to run. Depending on how many opportunities there are to run the `populate` fragments within a **Scenario**, the framework may run all of them together, or space them apart to run individually as the upgrade proceeds.
+
+`validate()` returns a testdrive fragment that validates that Mz is operating correctly with respect to any stuff that happened during population. The framework will start running this fragment at every opportunity after all the parts of the populating have completed.
+
+## An Example Explicit Check
+
+```
+@requires(initial_version > 001900)
+class KafkaSource(ExplicitCheck):
+    def populate(self) -> List[Td]:
+        return [
+            Td("$ kafka-create-topic + ... CREATE SOURCE ..."),
+            Td("$ kafka-ingest # Ingest first row"),
+            Td("$ kafka-ingest # Ingest second row"),
+        ]
```

Comment:
> is there anything stopping us from using .td files for complex populate's?

I am not sure I understand what you mean by "complex populate"? The intention is to use embedded testdrive commands to create kafka topics, define schemas and ingest data into topics. It would also be possible to call `Kgen` directly if needed.

Reply:
I mean "for complex `populate` methods, will it be possible to do something like:
```
    def populate(self) -> List[Td]:
        return tds_from_file("kafka-upsert.td")
```

where the file is local to the mzcompose workdir?

Reply:
yes, that would be possible, but by default we should strive to keep the "populate" and "validate" td content together